rust/sleeper_core/src/lib.rs

```diff
@@ -104,50 +106,106 @@ pub struct CommonConfig<'a> {
     /// Ranges for each column to filter input files
     pub region: SleeperPartitionRegion<'a>,
     /// How output from operation should be returned
-    pub output: OperationOutput,
+    pub output: CompletionOptions,
     /// Iterator config. Filters, aggregators, etc.
     pub iterator_config: Option<String>,
 }
 
-impl CommonConfig<'_> {
-    /// Convert all Java "s3a" URLs in input and output to "s3."
-    pub fn sanitise_java_s3_urls(&mut self) {
-        self.input_files.iter_mut().for_each(|t| {
-            if t.scheme() == "s3a" {
-                let _ = t.set_scheme("s3");
-            }
-        });
-
-        if let OperationOutput::File {
-            output_file,
-            opts: _,
-        } = &mut self.output
-            && output_file.scheme() == "s3a"
-        {
-            let _ = output_file.set_scheme("s3");
+impl Default for CommonConfig<'_> {
+    fn default() -> Self {
+        Self {
+            aws_config: Option::default(),
+            input_files: Vec::default(),
+            input_files_sorted: true,
+            row_key_cols: Vec::default(),
+            sort_key_cols: Vec::default(),
+            region: SleeperPartitionRegion::default(),
+            output: CompletionOptions::default(),
+            iterator_config: Option::default(),
         }
     }
+}
 
-    /// Checks for simple configuration errors
+impl<'a> CommonConfig<'a> {
+    /// Creates a new configuration object.
     ///
     /// # Errors
-    /// It is an error for input paths to be empty or for a length
-    /// mismatch between row key columns length and number of ranges in
-    /// partition region.
-    pub fn validate(&self) -> Result<()> {
-        if self.input_files.is_empty() {
-            bail!("No input paths supplied");
-        }
-        if self.row_key_cols.len() != self.region.len() {
-            bail!(
-                "Length mismatch between row keys {} and partition region bounds {}",
-                self.row_key_cols.len(),
-                self.region.len()
-            );
+    /// The configuration must validate. Input files mustn't be empty
+    /// and the number of row key columns must match the number of region
+    /// dimensions.
+    #[allow(clippy::too_many_arguments)]
+    pub fn try_new(
+        aws_config: Option<AwsConfig>,
+        input_files: Vec<Url>,
+        input_files_sorted: bool,
+        row_key_cols: Vec<String>,
+        sort_key_cols: Vec<String>,
+        region: SleeperPartitionRegion<'a>,
+        output: CompletionOptions,
+        iterator_config: Option<String>,
+    ) -> Result<Self> {
```

Comment:
It looks like this is only sometimes called, as all the fields of the struct are still public. It's also a lot of parameters to pass without names (where you set them against the field names if you instantiate the struct directly).

Might it be worth separating a spec/options struct from CommonConfig, where the fields are public in the spec/options struct but private in CommonConfig? That way we could be sure it's instantiated in one way, and the validation will be called.

Reply:
Done.